src/layout/Table.js

```diff
@@ -0,0 +1,108 @@
+  state: observable({
+    expandedRows: [],
+  }),
+}))(
+  observer(({ data, columns, recordKey = 'key', state, ...props }) => (
+    <table {...props.tableAttrs}>
+      <thead>
+        <tr>
+          {F.mapIndexed(
+            (c, i) => (
+              <th key={`${c.field}${i}`}>
+                {F.callOrReturn(_.getOr(F.autoLabel(c.field), 'label', c))}
+              </th>
+            ),
+            columns
+          )}
+        </tr>
+      </thead>
+      <tbody>
+        {_.map(
+          x => (
+            <React.Fragment key={x[recordKey]}>
+              <tr {...x.rowAttrs} key={x[recordKey]}>
+                {F.mapIndexed(
+                  ({ field, display = x => x, details = {} }, i) => (
+                    <td
+                      key={`${field}${i}`}
+                      {...!_.isEmpty(details) && {
+                        style: {
+                          cursor: !_.isEmpty(details) ? 'pointer' : 'auto',
+                        },
+                        onClick() {
+                          let indexedField = `${field}${i}`
+                          let expandedRow = state.expandedRows.find(
+                            row => row.key === x[recordKey]
+                          )
+                          if (!expandedRow) {
+                            state.expandedRows.push({
+                              key: x[recordKey],
+                              record: x,
+                              field,
+                              indexedField,
+                              details,
+                            })
+                          } else if (
+                            expandedRow.indexedField !== indexedField
+                          ) {
+                            expandedRow.field = field
+                            expandedRow.indexedField = indexedField
+                            expandedRow.details = details
+                          } else {
+                            state.expandedRows.remove(expandedRow)
+                          }
+                        },
+                      }}
+                    >
+                      {_.getOr(
+                        display,
+                        `${
+                          state.expandedRows.find(
+                            row =>
+                              row.key === x[recordKey] &&
+                              row.indexedField === `${field}${i}`
+                          )
+                            ? 'collapse'
+                            : 'expand'
+                        }.display`,
+                        details
+                      )(_.get(field, x), x)}
+                    </td>
+                  ),
+                  columns
+                )}
+              </tr>
+              {/* See if there is a details component to render for the column value when row expanded */}
+              {_.flow(
+                key => state.expandedRows.find(row => row.key === key),
```

Comment:
I wonder this would be better if we store the information as needed in the exampleRow. My intuition is that this might be easier to understand if it looks something like:

```javascript
{expandedRows[key] && <expandedRows[key].Component />}
```

So that the expandedRows is not an array but an object where the keys are the row.keys, and `expandedRow.component` is assigned like:
```javascript
expandedRows[recordKey] = {
  // ... the other props, then
  Component: () => _.get('component', details) && <tr>
  <td colSpan={columns.length}>
    {details.component(x[field], x)}
  </td>
</tr>
```

Sorry for the pseudocode! Let me know what you think!

Reply:
I tried that at first but since it is a dynamic key observable object, my observer was not picking up the changes, what do you think? I know we had something like expandObservable in the marketplace repo...

Reply:
Yes! so mobx 4 has a `set` method `import { set } from 'mobx'` which lets you alter an existing observable!

Reply:
I switched it to use an observable map (which seems that get and set in mobx do anyway when you use them). It cleaned up nicely.